packages/flutter/lib/src/widgets/overlay.dart

```diff
@@ -997,3 +959,1336 @@ class _RenderTheatre extends RenderBox with ContainerRenderObjectMixin<RenderBox
+
+  final _RenderTheatre theatre;
+  final _OverlayEntryWidgetState overlayEntryWidgetState;
+
+  @override
+  bool updateShouldNotify(_RenderTheatreMarker oldWidget) {
+    return oldWidget.theatre != theatre
+        || oldWidget.overlayEntryWidgetState != overlayEntryWidgetState;
+  }
+
+  @override
+  InheritedElement createElement() => _RenderTheatreMarkerElement(this);
+
+  static _RenderTheatreMarker? of(BuildContext context, { required bool targetRootOverlay }) {
+    final _RenderTheatreMarkerElement? element = context.getElementForInheritedWidgetOfExactType<_RenderTheatreMarker>() as _RenderTheatreMarkerElement?;
+    if (element == null) {
+      return null;
+    }
+    final _RenderTheatreMarkerElement dependency = targetRootOverlay ? element._topRenderTheatre : element;
+    return context.dependOnInheritedElement(dependency) as _RenderTheatreMarker;
+  }
+}
+
+class _RenderTheatreMarkerElement extends InheritedElement {
+  _RenderTheatreMarkerElement(super.widget);
+
+  _RenderTheatreMarkerElement get _topRenderTheatre {
+    _RenderTheatreMarkerElement? ancestor;
+    visitAncestorElements((Element element) {
+      assert(ancestor == null);
+      ancestor = element.getElementForInheritedWidgetOfExactType<_RenderTheatreMarker>() as _RenderTheatreMarkerElement?;
+      return false;
+    });
+    return ancestor?._topRenderTheatre ?? this;
+  }
+}
+
+class _OverlayPortalState extends State<OverlayPortal> {
+  // The developer must call `show` to reveal the overlay so we can get a unique
+  // timestamp of the user interaction for sorting.
+  _OverlayEntryLocation? _location;
+  int? _showTimestamp;
+
+  @override
+  void initState() {
+    super.initState();
+    _setupController(widget.controller);
+  }
+
+  void _setupController(OverlayPortalController controller) {
+    assert(
+      controller._attachTarget == null || controller._attachTarget == this,
+      'Failed to attach $controller to $this. It is already attached to ${controller._attachTarget}'
+    );
+    final int? controllerTimeStamp = controller._showTimestamp;
+    final int? timeStamp = _showTimestamp;
+    if (timeStamp == null || (controllerTimeStamp != null && controllerTimeStamp > timeStamp)) {
+      _showTimestamp = controllerTimeStamp;
+    }
+    controller._showTimestamp = null;
+    controller._attachTarget = this;
+  }
+
+  @override
+  void didChangeDependencies() {
+    super.didChangeDependencies();
+    _location?._dispose();
+    _location = null;
+  }
+
+  @override
+  void didUpdateWidget(OverlayPortal oldWidget) {
+    super.didUpdateWidget(oldWidget);
+    if (oldWidget.controller != widget.controller) {
+      oldWidget.controller._attachTarget = null;
+      _setupController(widget.controller);
+    }
+  }
+
+  @override
+  void dispose() {
+    widget.controller._attachTarget = null;
+    super.dispose();
+  }
+
+  void show(int timestampe) {
+    assert(SchedulerBinding.instance.schedulerPhase != SchedulerPhase.persistentCallbacks);
```

Comment:
This should probably come with a friendly error message since it seems likely that developers will run into this assert?

Reply:
Wait, but this assert will also prevent developers from calling "show" in e.g. initState, is that expected? I thought some earlier docs said its ok to call show if the portal hasn't been inflated or something like that?

Reply:
It won't, the assert is only in the `state.show()` so it's ok to call `show` on the controller in `initState` (since the `build` method of that state apparently hasn't been called yet and the controller is yet to be attached to the state). But still it's not advisable, it would reflect the build timestamp instead of the timestamp of the user interaction that triggered the `show` call.